Update semantic_versioning.md (#2842)
nf-core/website: sites/docs/src/content/docs/guidelines/pipelines/requirements/semantic_versioning.md

    weight: 90
---

- Pipelines must be released with stable release tags.
- Releases must use GitHub releases and include a [detailed changelog](https://keepachangelog.com/en/1.0.0/) file.
- Release version tags must be numerical only (no `v` prefix) and should follow [semantic versioning](https://semver.org/) rules: `[major].[minor].[patch]`

**Versioning Examples**

Starting from release version `1.4.3`, bumping the version to:

> - **`1.4.4`**: A patch release for minor fixes such as bug corrections that do not modify how the user interacts with the pipeline.

Reviewer: ```suggestion
- **`1.4.4`**: A patch release for minor fixes such as bug corrections that do not modify how the user interacts with the pipeline. Examples:
    - Inserting a missing argument to a process command line
    - Updating a container build version of a tool that is otherwise the same version of the tool itself
```

> - **`1.5`**: A minor release that adds new features without changing existing functionality.

Reviewer: ```suggestion
- **`1.5`**: A minor release that adds new features without changing existing functionality. Examples:
    - Adding a new parameter that adds new functionality/options to an existing module
    - Add a new optional or default process that does not change the way the user interacts with previous steps of a pipeline 
```

> - **`2.0`**: A major release where inputs or results are no longer backwards compatible, or when you rename or deprecate a parameter, or change the input samplesheet specification. This does not require major changes in workflow functionality—alterations to the pipeline's 'interface' are sufficient to justify a new major version.

Reviewer: ```suggestion
- **`2.0`**: A major release where execution interaction, inputs or results are no longer backwards in terms of pipeline user interaction. Examples:
    - Renaming or deprecations of existing parameter(s)
    - Change to input sample sheet specification (adding, dropping, renaming of mandatory columns) or change the input samplesheet specification. 
    - Note that a major release DOES NOT necessarily require new functionality
```

    weight: 90
---

- Pipelines must be released with stable release tags.
- Releases must use GitHub releases and include a [detailed changelog](https://keepachangelog.com/en/1.0.0/) file.
- Release version tags must be numerical only (no `v` prefix) and should follow [semantic versioning](https://semver.org/) rules: `[major].[minor].[patch]`

**Versioning Examples**

Starting from release version `1.4.3`, bumping the version to:

- **`1.4.4`**: A patch release for minor fixes such as bug corrections that do not modify how the user interacts with the pipeline. Examples:
  - Inserting a missing argument to a process command line

>   - Updating a container build version of a tool that is otherwise the same version of the tool itself

Reviewer 2: ```suggestion
  - Updating a container build version of a tool that is otherwise the same version of the tool itself
  - Fixing the logic of the pipeline operators to ensure correct inputs for a process
```

> - **`1.5`**: A minor release that adds new features without changing existing functionality. Examples:

Reviewer 2: ```suggestion
- **`1.5.0`**: A minor release that adds new features without changing existing functionality. Examples:
```

  - Adding a new parameter that adds new functionality/options to an existing module
  - Add a new optional or default process that does not change the way the user interacts with previous steps of a pipeline

> - **`2.0`**: A major release where execution interaction, inputs or results are no longer backwards in terms of pipeline user interaction. Examples:

Reviewer 2: ```suggestion
- **`2.0.0`**: A major release where execution interaction, inputs or result structures are no longer backwards in terms of pipeline user interaction. Examples:
```

Reviewer: Ah that's a good one suggestion, I can understand now better how that could be braking!

  - Renaming or deprecations of existing parameter(s)

>   - Change to input sample sheet specification (adding, dropping, renaming of mandatory columns) or change the input samplesheet specification.

Reviewer 2: ```suggestion
  - Change to input sample sheet specification (adding, dropping, renaming of mandatory columns) or change the input samplesheet specification
```

    weight: 90
---

- Pipelines must be released with stable release tags.
- Releases must use GitHub releases and include a [detailed changelog](https://keepachangelog.com/en/1.0.0/) file.
- Release version tags must be numerical only (no `v` prefix) and should follow [semantic versioning](https://semver.org/) rules: `[major].[minor].[patch]`

**Versioning Examples**

Starting from release version `1.4.3`, bumping the version to:

> - **`1.4.4`**: A patch release for minor fixes such as bug corrections that do not modify how the user interacts with the pipeline. Examples:
>   - Inserting a missing argument to a process command line
>   - Updating a container build version of a tool that is otherwise the same version of the tool itself

Reviewer 3: What about having a checklist insteaad, something such as:
```suggestion
## Release type checklist

If your release matches any point in the checklists below, classify it according to the highest applicable category in this order: major, minor, then patch. A single match in a higher category overrides all lower categories.

### Major Release (e.g., 1.0.0 to 2.0.0)

- [ ] Changes that are not backwards compatible in terms of pipeline user interaction
- [ ] Renaming or deprecation of existing parameter(s)
- [ ] Changes to input sample sheet specification (adding, dropping, renaming of mandatory columns)
- [ ] Significant changes to the pipeline's core functionality or workflow
- [ ] Major changes to output formats or directory structures
- [ ] Removal of previously supported features or tools
- [ ] Changes that require users to modify their existing configuration files or scripts
- [ ] Upgrades to dependencies that introduce breaking changes

## Minor Release (e.g., 1.4.0 to 1.5.0)

- [ ] Addition of new features without changing existing functionality
- [ ] Adding a new parameter that introduces new functionality/options to an existing module
- [ ] Adding a new optional or default process that doesn't change user interaction with previous steps
- [ ] Introducing new output files or metrics without altering existing ones
- [ ] Adding support for new input formats or data types
- [ ] Implementing performance improvements that don't affect the pipeline's interface
- [ ] Expanding documentation or adding new examples without changing core functionality
- [ ] Adding new optional dependencies or tools

## Patch Release (e.g., 1.4.3 to 1.4.4)

- [ ] Minor fixes such as bug corrections that do not modify how the user interacts with the pipeline
- [ ] Inserting a missing argument to a process command line
- [ ] Updating a container build version of a tool that is otherwise the same version of the tool itself
- [ ] Fixing the logic of pipeline operators to ensure correct inputs for a process
- [ ] Correcting typos in documentation or code comments
- [ ] Updating metadata or non-functional elements (e.g., README, CONTRIBUTING files)
- [ ] Making minor adjustments to log messages or error handling
- [ ] Applying small optimizations that don't affect functionality or user interaction

When deciding on the release type, it should. be considered the impact on users and whether they need to make changes to their existing workflows or configurations. If in doubt, it's generally safer to opt for a higher version increment to clearly communicate the extent of changes to users.
```

Reviewer 3: Yes, that is why I included the message above the checklist. Anyway, my point is more about how having a list should make it easier for developers to see which category the release corresponds to. So a regular list would be also ok.

Reviewer 3: Modified my comment accordingly ☝️

Reviewer 4: ah, didn't read everything, sorry, yes, actually a nice idea

Reviewer 2: I like this idea of a list

Author: OK, I'm going to merge this suggestion and we can tweak if necessary from there.

  - Fixing the logic of the pipeline operators to ensure correct inputs for a process
- **`1.5.0`**: A minor release that adds new features without changing existing functionality. Examples:
  - Adding a new parameter that adds new functionality/options to an existing module
  - Add a new optional or default process that does not change the way the user interacts with previous steps of a pipeline

> - **`2.0.0`**: A major release where execution interaction, inputs or result structures are no longer backwards in terms of pipeline user interaction. Examples:

Author: ```suggestion
- **`2.0.0`**: A major release where execution interaction, inputs or result structures are no longer backwards compatible in terms of pipeline user interaction. Examples:
```

  - Renaming or deprecations of existing parameter(s)
  - Change to input sample sheet specification (adding, dropping, renaming of mandatory columns) or change the input samplesheet specification

>   - Note that a major release DOES NOT necessarily require new functionality

Author: ```suggestion
  - Note that a major release DOES NOT necessarily require new functionality
  - It's expected that regular breaking changes might occur in early stages of pipeline development prior to version 1.0.0, so the criteria don't apply there. 
```

I think we need something like this

Reviewer 4: we discourage to do pre-releases in nf-core, so this is not be needed, I think

Author: So the first release should always be V1.0.0? We should probably make that explicit somewhere (maybe not here)

Reviewer 4: I thought we had that somewhere, but don't find it at the moment... let's add it then here and in the "first release" tutorial

Author: Done!

- The first release of a pipline should be version `1.0.0` (we strongly discourage pre-releases).
- Release version tags must be numerical only (no `v` prefix) and should follow [semantic versioning](https://semver.org/) rules: `[major].[minor].[patch]`

## Semantic Versioning Criteria

When deciding on the release type, consider the impact on users and whether they need to make changes to their existing workflows or configurations. If in doubt, opt for a higher version increment to clearly communicate the extent of changes to users.

Classify your release according to the highest applicable category in this order: **Major**, **Minor**, then **Patch**. A single match in a higher category overrides all lower categories.

### Major Release (e.g., 1.4.3 to 2.0.0)

Changes that are **not backwards compatible** and may require users to adjust their existing workflows or configurations.

- **Breaking Changes**:
  - Modifications to pipeline user interaction that are not backwards compatible.
  - Renaming or deprecation of existing parameter(s).
  - Changes to input sample sheet specifications (adding, dropping, or renaming mandatory columns).

>   - Major alterations to output formats or directory structures.

Reviewer: ```suggestion
  - Major alterations to existing output formats (e.g., BAM to CRAM, or changed TSV headers names) or directory structures.
```

  - Removal of previously supported features or tools.

>   - Upgrades to dependencies that introduce breaking changes.

Reviewer: What does that mean exactly? Examples are important here

- **Significant Functional Changes**:

>   - Overhauls to the pipeline's core functionality or workflow.

Reviewer: Need examples

Author: Let's just take this out, I think it will be covered by things in 'breaking changes'.

> **Note:** A major release does not necessarily require new functionality; it can result solely from breaking changes.

Reviewer: ```suggestion
:::note
A major release does not necessarily require new functionality; it can result solely from breaking changes.
:::
```

### Minor Release (e.g., 1.4.3 to 1.5.0)

Introduction of new features that are **backwards compatible** and do not alter existing functionality.

- **Feature Additions**:

>   - Adding new parameters that introduce additional functionality or options to existing modules.

Reviewer: ```suggestion
  - Adding new parameters that introduce additional functionality or options to existing modules (without fundamentally changing output file types or structure).
```

  - Incorporating new optional or default processes without changing user interaction with previous steps.
  - Introducing new output files or metrics without modifying existing ones.

>   - Supporting new input formats or data types.

Reviewer: ```suggestion
  - Supporting new input formats or data types without removing support for old ones.
```

- **Improvements**:

>   - Implementing performance enhancements that don't affect the pipeline's interface.

Reviewer: ```suggestion
  - Implementing performance enhancements that don't affect the pipeline's interface (e.g. refined default memory or CPU requirements).
```

  - Expanding documentation or adding new examples without altering core functionality.

>   - Adding new optional dependencies or tools.

Reviewer: Isn't that the same as point2?

### Patch Release (e.g., 1.4.3 to 1.4.4)

Minor changes that do not affect user interaction or existing functionality.

- **Bug Fixes**:
  - Correcting minor bugs that do not change how the user interacts with the pipeline.
  - Fixing the logic of pipeline operators to ensure correct inputs for a process.

>   - Inserting missing arguments into process command lines.

Reviewer: ```suggestion
  - Inserting missing arguments into process command lines that should've existed before (but without modifying output file types/output file content structure).
```

- **Updates and Corrections**:

>   - Updating a container build version of a tool without changing the tool's version itself.

Reviewer: ```suggestion
  - Updating a container build version of a tool without changing the tool's version itself (e.g. so it can execute correctly).
```

## Semantic Versioning Criteria

When deciding on the release type, consider the impact on users and whether they need to make changes to their existing workflows or configurations. If in doubt, opt for a higher version increment to clearly communicate the extent of changes to users.

Classify your release according to the highest applicable category in this order: **Major**, **Minor**, then **Patch**. A single match in a higher category overrides all lower categories.

### Major Release (e.g., 1.4.3 to 2.0.0)

Changes that are **not backwards compatible** and may require users to adjust their existing workflows or configurations.

- **Breaking Changes**:
  - Modifications to pipeline user interaction that are not backwards compatible.
  - Renaming or deprecation of existing parameter(s).
  - Changes to input sample sheet specifications (adding, dropping, or renaming mandatory columns).
  - Major alterations to existing output formats (e.g., BAM to CRAM, or changed TSV headers names) or directory structures.
  - Removal of previously supported features or tools.

>   - Upgrades to dependencies that introduce breaking changes.

Author: ```suggestion
  - Upgrades to dependencies that introduce breaking changes. For example if the new version of the tool requires additional parameters, removes parameters passed by the pipeline, or changes its outputs in a way that necessitates changes to module output definitions. 
```

> - **Significant Functional Changes**:
>   - Overhauls to the pipeline's core functionality or workflow.

Author: ```suggestion
```

  - Removal of previously supported features or tools.
  - Upgrades to dependencies that introduce breaking changes.
- **Significant Functional Changes**:
  - Overhauls to the pipeline's core functionality or workflow.

:::note
A major release does not necessarily require new functionality; it can result solely from breaking changes.
:::

### Minor Release (e.g., 1.4.3 to 1.5.0)

Introduction of new features that are **backwards compatible** and do not alter existing functionality.

- **Feature Additions**:
  - Adding new parameters that introduce additional functionality or options to existing modules (without fundamentally changing output file types or structure).
  - Incorporating new optional or default processes without changing user interaction with previous steps.
  - Introducing new output files or metrics without modifying existing ones.
  - Supporting new input formats or data types without removing support for old ones.
- **Improvements**:
  - Implementing performance enhancements that don't affect the pipeline's interface (e.g. refined default memory or CPU requirements).
  - Expanding documentation or adding new examples without altering core functionality.

>   - Adding new optional dependencies or tools.

Author: ```suggestion
```
